diff --git a/optimum/exporters/openvino/__main__.py b/optimum/exporters/openvino/__main__.py
index 18f650c..ddae47c 100644
--- a/optimum/exporters/openvino/__main__.py
+++ b/optimum/exporters/openvino/__main__.py
@@ -47,6 +47,7 @@ logger = logging.getLogger(__name__)
 def main_export(
     model_name_or_path: str,
     output: Union[str, Path],
+    adapter_path: str = None,
     task: str = "auto",
     device: str = "cpu",
     framework: Optional[str] = None,
@@ -240,6 +241,11 @@ def main_export(
         **loading_kwargs,
     )
 
+    if adapter_path is not None:
+        from peft import PeftModel
+        PeftModel.from_pretrained(model, adapter_path)
+        logger.info(f"Adapters loaded from {adapter_path} are attached")
+
     needs_pad_token_id = task == "text-classification" and getattr(model.config, "pad_token_id", None) is None
 
     if needs_pad_token_id:
diff --git a/optimum/intel/openvino/modeling_decoder.py b/optimum/intel/openvino/modeling_decoder.py
index 0816557..a41bf0c 100644
--- a/optimum/intel/openvino/modeling_decoder.py
+++ b/optimum/intel/openvino/modeling_decoder.py
@@ -262,9 +262,11 @@ class OVBaseDecoderModel(OVModel):
             compression_option = "fp32"
 
         stateful = kwargs.pop("stateful", ensure_stateful_is_available(warn=False) and use_cache)
+        adapter_path = kwargs.pop("adapter_path", None)
         main_export(
             model_name_or_path=model_id,
             output=save_dir_path,
+            adapter_path=adapter_path,
             task=task,
             subfolder=subfolder,
             revision=revision,
